Feature/1724 typed hashes

davidnich opened the pull request.

davidnich committed:
refs #1724 started working on typed hashes

davidnich committed:
Merge remote-tracking branch 'origin/develop' into feature/1724_types_hashes

davidnich committed:
refs #1724 more progress on typed hash implementation

davidnich committed:
Merge remote-tracking branch 'origin/develop' into feature/1724_types_hashes

davidnich committed:
Merge remote-tracking branch 'origin/develop' into feature/1724_types_hashes

davidnich committed:
refs #1724 more progress on typed hash support

davidnich committed:
refs #1724 started rewriting internal type system to allow for complex types and a new typed hash type

davidnich committed:
Merge remote-tracking branch 'origin/develop' into feature/1724_typed_hashes

davidnich committed:
Merge remote-tracking branch 'origin/develop' into feature/1724_typed_hashes

davidnich committed:
refs #1724 more progress on a new internal type system

davidnich committed:
refs #1724 create replacement classes for all existing type classes

davidnich committed:
Merge remote-tracking branch 'origin/develop' into feature/1724_typed_hashes

davidnich committed:
refs #1724 partially working new internal type system

davidnich committed:
refs #1724 more progress on the new internal type implementation

davidnich committed:
refs #1724 fixed class cimpatibility issues with injected classes, necessary due to improved type support

davidnich committed:
refs #2001 fixed date/time parse-time type reporting, re-enabled tests in types.qtest

davidnich committed:
refs #1724 fixed type matching with classes

davidnich committed:
Merge remote-tracking branch 'origin/develop' into feature/1724_typed_hashes

davidnich committed:
refs #1724 implement support for subtype declaration lists

davidnich committed:
refs #1724 working hashdecl initialization and casting

davidnich committed:
refs #1724 more progress on typed hash implementation, working key type enforcement and casts

davidnich committed:
refs #1724 type matching fixes

davidnich committed:
refs #1724 fixed hash lvalue handling

davidnich committed:
refs #1724 another typed hash lvalue fix

davidnich committed:
refs #1724 fixes, updates + initial doc updates

davidnich committed:
refs #1724 doc updates

davidnich committed:
refs #1724 added missing doc file

davidnich committed:
refs #1724 fixed non-single-compilation-unit builds

davidnich committed:
refs #1724 major cmake doc generation corrections

davidnich committed:
refs #1724 minor doc updates

davidnich committed:
refs #1724 added missing file

davidnich committed:
refs #1724 fixed complex type declaration scanning with namespace justification

davidnich committed:
refs #1724 many hashdecl fixes, implemented support for hashes with an explicit value type

davidnich committed:
Merge remote-tracking branch 'origin/develop' into feature/1724_typed_hashes

davidnich committed:
refs #1724 minor internal fixes, added type-safety to ConnectionProvider hashes, updated qdx and Qdx.qm to support hashdecls in user code

davidnich committed:
refs #1724 more progress on types and complex type integration

davidnich committed:
refs #1724 added missing file

davidnich committed:
refs #1724 more complex type/hashdecl fixes

davidnich committed:
Merge remote-tracking branch 'origin/develop' into feature/1724_typed_hashes

davidnich committed:
refs #1724 updated hash() to convert type-safe values to untyped hashes; updated docs

davidnich committed:
refs #1724 fixed a memory leak in parsing hashdecl declarations

davidnich committed:
refs #1724 minor corrects and started using complex types in SqlUtil

davidnich committed:
refs #1724 fixed a memory leak in hashdecl parsing

davidnich committed:
refs #1724 fixed another parser memory bug related to hashdecl processing; more hashdecl integration in standard modules

davidnich committed:
refs #1724 many type fixes, complex hashes now support runtime type folding

davidnich committed:
refs #1724 implement support for checking if an lvalue assignment might be subject to runtime type folding

davidnich committed:
refs #1724 many parse and runtime type handling fixes

davidnich committed:
refs #1724 implmented qpp support for hashdecls; added hashdecls to qore; fixed Program API management regarding hashdecls

davidnich committed:
refs #1724 fixed a bug introduced in the last commit in a new internal hash API

davidnich committed:
refs #1724 misc hashdecl fixes

davidnich committed:
refs #1724 various type matching fixes, overload tests for complex hash types

davidnich committed:
refs #1724 added new pseudo-methods to expose complex type information at runtime

davidnich committed:
refs #1724 fixed an ambiguous constructor in TableMapper::InboundTableMapperIterator()

davidnich committed:
refs #1724 more type-safe hash fixes

davidnich committed:
refs #1724 fixed hash key APIs

davidnich committed:
refs #1724 started adding type safety to lists, added a test (currently failing) for complex hashes

davidnich committed:
refs #1724 more complex type fixes

davidnich committed:
refs #1724 initial support for typed lists

davidnich committed:
refs #1724 fixed memory leaks introduced in recent commits

davidnich committed:
refs #1724 more progress on strongly-typed lists

davidnich committed:
refs #1724 more progress on type-safe lists

davidnich committed:
refs #1724 more progress on typed lists, must better parse type detection

davidnich committed:
refs #1724 fixed broken list handling

davidnich committed:
refs #1724 doc updates, fixed parse-time type checking when casting to a plain list or hash

davidnich committed:
refs #1724 fix typo in docs

davidnich committed:
refs #1724 really fixed the same typo for the second time

davidnich committed:
refs #1724 added qpp support for complex hashes and lists

davidnich committed:
refs #1724 added support to qpp for complex types, added type-safety to the thread stack APIs, fixed base class constructor argument error reporting (again)

davidnich committed:
refs #1724 doc updates

davidnich committed:
refs #1724 use ExceptionInfo hashdecl has exception argument type

davidnich committed:
Merge remote-tracking branch 'origin/develop' into feature/1724_typed_hashes

davidnich committed:
refs #1724 fixed a memory leak in the try/catch statement introduced in a recent commit

davidnich committed:
refs #1724 fixed a memory leak in new complex list expressions

davidnich committed:
refs #1724 fixed issues with argument optimizations and non-scu builds

davidnich committed:
refs #1724 fixed compile warnings

davidnich committed:
refs #1724 reverted unnecessary changes to params.h

davidnich committed:
refs #1724 fixed RestHandler.qtest to run with no serialization modules present

davidnich committed:
refs #1724 fixed a bug introduced in a recent comment where an address was stored of a temporary variable

davidnich committed:
refs #1724 implemented support for consise complex type declaration & initialization for hashdecls and class members

davidnich committed:
refs #1724 added a test for complex initializtion in hashdecls

davidnich committed:
refs #1724 fixed RestHandler.qtest with serialization modules in place and fixed memory issues introduced in the last commit

davidnich committed:
refs #1724 fixed <hash>::getKey() return type value, added simple list tests, cannot fix more complex parse-time return values for pseudo-methods with the current type system - we would need a template- or generics-based implementation like:

template <typename T>
list<T> <hash<string, T>> <hash>::getValues() {...}

template <typename T>
*T <hash<string, T>> <hash>::getFirstValue() {...}

this is currently not possible, so we have to stick to plain types for now, however the pseudo-methods base value type info was added to MethodCallNode for this purpose for future use

davidnich committed:
refs #1724 fixed string split and regular expression extract return types in all functions and pseudo-methods, added tests

davidnich committed:
refs #1724 implemented support for strong type support for implicit arguments where possible

davidnich committed:
refs #1724 fixed a memory error in the internal scan of complex types

davidnich committed:
refs #1724 fixed angle bracket type parsing with multiple types on the same line

davidnich committed:
refs #1724 fixed a bug in exception handling introduced in this branch; added a test

davidnich committed:
refs #1724 added missing include for non-scu builds

davidnich committed:
refs #1724 readded accidentally deleted public symbol

davidnich committed:
Merge branch 'develop' into feature/1724_typed_hashes

Commit:
Typos, whitespace, docs fixes

omusil24 commented:
@davidnich

1) Why have special class `TypedHashDeclHolder` instead of using `std::unique_ptr`?

2) lib/HashDeclList.cpp:69 - why not return -1 as usual?

3) lib/QoreListNode.cpp:408, 414 - UNSAFE methods... When will they be removed?

4) lib/parser.ypp:436, 609, 759 - why not use enums instead of defines?

5) lib/ql_lib.qpp:121 - missing end of comment: "for example <what>"

6) qlib/HttpServerUtil.qm::242, 266 - why `close` keys don't have the same type?

7) include/qore/intern/QoreParseListNode.h: 52->55->68 -  `lvec` seems to be getting filled with "null" or possibly even garbage locations...

Still have to check `QoreTypeInfo.h` and `QoreTypeInfo.cpp`...

davidnich committed:
refs #1724 corrected return value (1 -> -1)

davidnich committed:
refs #1724 aligned types of complemtary hashdecls in HttpServerUtil.qm

davidnich committed:
refs #1724 fixed location copying in the QoreParseListNode copy constructor

davidnich commented:
@omusil24 
>  1. Why have special class TypedHashDeclHolder instead of using std::unique_ptr?

for the same reason QoreClass has it - it's reference counted and therefore has a non-public destructor.  `TypedHashDeclHolder` makes it easy and safe to use in module code for example.

> 2. lib/HashDeclList.cpp:69 - why not return -1 as usual?

was a typo - now fixed

> 3. lib/QoreListNode.cpp:408, 414 - UNSAFE methods... When will they be removed?

when the ABI changes - in the next major release of Qore - 0.9; I would like to do this after 0.8.13 - 0.9 should have a new ABI and should eliminate QoreBigIntNode and QoreFloatNode and others in favor or QoreValue as  the main holder; etc.  If we remove these unsafe functions now then we make an ABI change.

> 4. lib/parser.ypp:436, 609, 759 - why not use enums instead of defines?

no problem if you'd like to change it to use an enum - sounds fine

> 5. lib/ql_lib.qpp:121 - missing end of comment: "for example "

that's intentional and makes sense to me at least.  It's OK to end a sentence with "for example", meaning that the previous statement was an example.

> 6. qlib/HttpServerUtil.qm::242, 266 - why close keys don't have the same type?

fixed

> 7. qlib/HttpServerUtil.qm::242, 266 - why close keys don't have the same type?

fixed

davidnich committed:
refs #1724 implemented a runtime check for recursive hashdecl initialization

Commit:
Merge branch 'develop' into feature/1724_typed_hashes
